graveler/committed/merge_iterator.go

```diff
@@ -0,0 +1,109 @@
+package committed
+
+import (
+	"bytes"
+
+	"github.com/treeverse/lakefs/graveler"
+)
+
+type mergeIterator struct {
+	diffIt graveler.DiffIterator
+	val    *graveler.ValueRecord
+	base   Iterator
+	err    error
+}
+
+// NewMergeIterator accepts an iterator describing a diff from theirs to ours.
+// It returns a ValueIterator with the changes to perform on theirs, in order to merge ours into it,
+// relative to base as the merge base.
+// The iterator will return ErrConflictFound when it reaches a conflict.
+func NewMergeIterator(diffTheirsToOurs graveler.DiffIterator, base Iterator) (*mergeIterator, error) {
+	return &mergeIterator{diffIt: diffTheirsToOurs, base: base}, nil
+}
+
+func (d *mergeIterator) valueFromBase(key graveler.Key) *graveler.ValueRecord {
+	d.base.SeekGE(key)
+	var val *graveler.ValueRecord
+	for d.base.Next() && val == nil {
```

Comment:
`val == nil` will always be true here because you just declared it and haven't assigned to it yet.

Reply:
but it's a loop